File: app/src/main/java/com/appchamp/wordchunks/packs/PacksActivity.java
package com.appchamp.wordchunks.packs;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.appchamp.wordchunks.R;
import com.appchamp.wordchunks.data.DatabaseHelperRealm;
import com.appchamp.wordchunks.levels.LevelsActivity;
import com.appchamp.wordchunks.models.realm.Pack;
import com.appchamp.wordchunks.util.AnimUtils;

import java.util.List;

import io.realm.Realm;
import me.everything.android.ui.overscroll.OverScrollDecoratorHelper;

import static com.appchamp.wordchunks.util.Constants.EXTRA_PACK_ID;


public class PacksActivity extends AppCompatActivity {

    private Realm realm;

    private DatabaseHelperRealm db;

    private List<Pack> packs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_packs);

        RecyclerView rvPacks = (RecyclerView) findViewById(R.id.rvPacks);

        db = new DatabaseHelperRealm();

        realm = Realm.getDefaultInstance();

        // This is the RecyclerView adapter which will display the list of packs
        packs = db.findAllPacks(realm);

        PacksAdapter adapter = new PacksAdapter(packs);
        rvPacks.setAdapter(adapter);
        rvPacks.setLayoutManager(new LinearLayoutManager(PacksActivity.this));
        rvPacks.setHasFixedSize(true);

        adapter.setOnItemClickListener((view, position) -> {
            Pack clickedPack = adapter.getItem(position);
            startLevelsActivity(clickedPack.getId());
        });

        OverScrollDecoratorHelper.setUpOverScroll(
                rvPacks, OverScrollDecoratorHelper.ORIENTATION_VERTICAL);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realm.close(); // Remember to close Realm when done.
    }

    public void onBackArrowClick(View v) {
        super.onBackPressed();
        AnimUtils.startAnimationFadeIn(PacksActivity.this, v);
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
    }

    /**
     * Up navigation. Navigates from PacksActivity to LevelsActivity passing packId by the Intent.
     * @param packId pack id to be passed by the Intent.
     */
    private void startLevelsActivity(String packId) {
        Intent intent = new Intent(PacksActivity.this, LevelsActivity.class);
        intent.putExtra(EXTRA_PACK_ID, packId);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
        finish();
        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left);
    }
}

File: app/src/main/java/com/appchamp/wordchunks/ui/packslevels/PacksActivity.kt
package com.appchamp.wordchunks.ui.packslevels

import android.content.Context
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import com.appchamp.wordchunks.R
import com.appchamp.wordchunks.models.realm.Pack
import com.appchamp.wordchunks.util.Constants.EXTRA_PACK_ID
import com.appchamp.wordchunks.util.queryAll
import kotlinx.android.synthetic.main.act_packs_levels.*
import me.everything.android.ui.overscroll.OverScrollDecoratorHelper
import org.jetbrains.anko.clearTop
import org.jetbrains.anko.intentFor
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper


class PacksActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.act_packs_levels)

        llPacksLevels.setBackgroundResource(R.drawable.gradient_packs)
        tvTitle.text = getString(R.string.title_select_pack)
        imgBackArrow.setOnClickListener { onBackPressed() }
        rvList.layoutManager = LinearLayoutManager(this)
        rvList.setHasFixedSize(true)

        OverScrollDecoratorHelper
                .setUpOverScroll(rvList, OverScrollDecoratorHelper.ORIENTATION_VERTICAL)
    }

    override fun onResume() {
        super.onResume()
        loadPacks()
    }

    private fun loadPacks() {
        val packs = Pack().queryAll()
        val adapter = PacksLevelsAdapter(packs) { startLevelsActivity(it.id) }
        rvList.adapter = adapter
    }

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase))
    }

    override fun onBackPressed() {
        super.onBackPressed()
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right)
    }

    /**
     * Navigates from PacksActivity to LevelsActivity passing packId by the Intent.
     * @param packId pack id to be passed by the Intent.
     */
    private fun startLevelsActivity(packId: String?) {
        startActivity(intentFor<LevelsActivity>(EXTRA_PACK_ID to packId).clearTop())
        finish()
        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left)
    }
}